libnmstate/state.py

```diff
@@ -935,3 +953,52 @@ def _remove_route_rule_default_values(rule):
     if rule.get(RouteRule.ROUTE_TABLE) == RouteRule.USE_DEFAULT_ROUTE_TABLE:
         del rule[RouteRule.ROUTE_TABLE]
     return rule
+
+
+def merge_linux_bridge_ports(desired_iface_state, current_iface_state):
+    """
+    Given a linux bridge desired state, and it's current state, merges
+    those together.
+
+    This extension of the interface merging mechanism simplifies the user's
+    life in scenarios where the user wants to partially update the bridge's
+    configuration - e.g. update only the bridge's port STP configuration -
+    since it enables the user to simply specify the updated values rather
+    than the full current state + the updated value.
+    """
+    desired_ports = desired_iface_state.get(
+        LinuxBridge.CONFIG_SUBTREE, {}
+    ).get(LinuxBridge.PORT_SUBTREE)
+    current_ports = current_iface_state.get(
+        LinuxBridge.CONFIG_SUBTREE, {}
+    ).get(LinuxBridge.PORT_SUBTREE, [])
+
+    if desired_ports is None and current_ports:
```

Comment:
As I commented before, I do not really like this complexity from this point down.
I want to check if we can simplify it.

- If the port subtree was not mentioned in the desired state, I would expect it to be merged from the current state by the original `merge_interfaces` processing. How come is it not?
- If the port subtree is mentioned in the desired state as an explicit empty list, I would expect it to overwrite the current state as part of the original merge as well.
- We are left with merging individual dicts, for the ones that appear in both the desired and the current. This is exactly what the `merge_interfaces` do at the interface list level, therefore doing the same here on the port list level could work as well. Please help me understand if it cannot.

I have also posted #913 which suggests to treat an explicit empty dict as an explicit request from the user/client to initialize/cleanup a subtree in the state. This fits the idea discussed here to remove the vlan subtree.

Reply:
I've rebased the tree with your PR - removing the interface merging commit - and it does not work.

It fails to update the desired state with the current state on the following tests:
  - https://github.com/nmstate/nmstate/pull/888/files#diff-d74d6ba3d078c4946b2ef63a5e16130fR1409
    - failed to merge vlan configuration from port1 into the desired state - check [0]
  - https://github.com/nmstate/nmstate/pull/888/files#diff-d74d6ba3d078c4946b2ef63a5e16130fR1530
    - failed to merge the vlan configuration from the port - check [1]

Seems the state is not being correctly passed from the current into the desired state.

[0] - https://paste.centos.org/view/3759b702
[1] - https://paste.centos.org/view/38ff380b

These pastebins will expire after 1 day.